diff --git a/src/jit/dump.lua b/src/jit/dump.lua
index 18e7a4b7..6a744c95 100644
--- a/src/jit/dump.lua
+++ b/src/jit/dump.lua
@@ -43,6 +43,7 @@
 --    T  Plain text output.
 --    A  ANSI-colored text output
 --    H  Colorized HTML + CSS output.
+--    J  JSON output.
 --
 -- The default output format is plain text. It's set to ANSI-colored text
 -- if the COLORTERM variable is set. Note: this is independent of any output
@@ -709,6 +710,10 @@ local function dumpon(opt, outfile)
     colorize = colorize_html
     irtype = irtype_html
     out:write(header_html)
+  elseif colormode == "J" then
+    colorize = colorize_text
+    irtype = irtype_text
+    out:write(header_html)
   else
     colorize = colorize_text
     irtype = irtype_text
