From f70b0bc12097510287e15517cdc6d1dcec9353b0 Mon Sep 17 00:00:00 2001
From: Churkin Aleksey <lekhachurkin@gmail.com>
Date: Sun, 9 Nov 2025 16:42:22 +0300
Subject: [PATCH] patch some internal to expose snapshot pc in bytecode

---
 src/lib_jit.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/src/lib_jit.c b/src/lib_jit.c
index 9f870f68..7fc37a63 100644
--- a/src/lib_jit.c
+++ b/src/lib_jit.c
@@ -288,6 +288,31 @@ static const char *const jit_trlinkname[] = {
   "interpreter", "return", "stitch"
 };
 
+/* local info = jit.util.snappc(tr) */
+LJLIB_CF(jit_util_snappc)
+{
+  int snapno = lj_lib_checkint(L, 2);
+
+  GCtrace *T = jit_checktrace(L);
+  if (snapno < 0 || snapno >= T->nsnap) {
+    lua_pushnil(L);  // Invalid snapshot number
+    return 1;
+  }
+  SnapShot *snap = &T->snap[snapno];
+  SnapEntry *map = &T->snapmap[snap->mapofs];
+  const BCIns *pc = snap_pc(&map[snap->nent]);
+  GCproto *pt = (GCproto *)gcref(T->startpt);
+  if (!pt) {
+    lua_pushnil(L);  // No prototype available
+    return 1;
+  }
+  BCIns *bc_base = proto_bc(pt);
+  int offset = (int)(pc - bc_base);  // Instruction offset
+  // Return as Lua number (instruction offset)
+  lua_pushinteger(L, offset);
+  return 1;
+}
+
 /* local info = jit.util.traceinfo(tr) */
 LJLIB_CF(jit_util_traceinfo)
 {
-- 
2.43.0

